Experimental support for CRB 15000-12/1.27

mmmarcopalma opened the pull request:
Extending support for CRB 15000 to model 12-1.27 (GoFa 12 kg). 

Includes updates of: 

- `/config`, `/launch`, `/urdf` directories 
- .stl visual and collision mesh geometry 
   (link-3, link-5 and link-6 are referenced to `meshes/crb15000_5_95`)

mmmarcopalma committed:
Added support for CRB 15000-12

mmmarcopalma commented:
/cc @gonzalocasas

gavanderhoorn commented:
Thanks for the PR.

Some high-level comments:

- could you remove all the `desktop.ini` files please?
- could none of the meshes of the other variant be reused?
- please add a `roslaunch` test for the new variant
- please update the description in the manifest with the details about the new variant you added
- and if you'd like: please add yourself as an `author`, seeing as you added a new variant

mmmarcopalma commented:
Thanks, I'll follow up on your comments.

> * could none of the meshes of the other variant be reused?

I could re-use 3+3 meshes (links 3-5-6). The ones I uploaded are exact copies of the existing ones.   

I'm not fluent yet in git, apologies for missing on some best-practice.

gavanderhoorn commented:
> > * could none of the meshes of the other variant be reused?
> 
> I could re-use 3+3 meshes (links 3-5-6). The ones I uploaded are exact copies of the existing ones.

Ok, so this is perhaps due to me not being clear by what I meant by *reuse*.

A `xacro:macro` can refer to files 'anywhere', so you don't have to copy meshes if they are identical to files for other variants.

Just have your `<mesh filename="..."/>` point to the file you'd like to reuse from the other variant.

No need to copy anything.

mmmarcopalma committed:
Added support for CRB 15000-12

mmmarcopalma committed:
Update package.xml

mmmarcopalma commented:
I added a commit with the requested changes. 
 
> * and if you'd like: please add yourself as an `author`, seeing as you added a new variant

Thanks, I added my name in the Package.xml

mmmarcopalma commented:
Hello @gavanderhoorn, is there any news regarding this request? 
Thanks!

gavanderhoorn commented on abb_crb15000_support/urdf/crb15000_12_127_macro.xacro:
```diff
@@ -0,0 +1,169 @@
+      <axis xyz="0 0 1"/>
+      <parent link="${prefix}base_link"/>
+      <child link="${prefix}link_1"/>
+      <limit effort="0" lower="${radians(-180)}" upper="${radians(180)}" velocity="${radians(125)}"/>
```
Could you clarify which specific version of the `12/1.27` you're modelling here? According to the documentation I've found (`Product manual, CRB 15000`, `3HAC077389-001`, `Rev L`), there is an option for +-270 degrees as well:

![image](https://github.com/user-attachments/assets/e067b8e3-ba68-4f0c-bef0-669fae7bd345)

gavanderhoorn replied:
The joint velocity limit also seems off:

![image](https://github.com/user-attachments/assets/3af25e28-eec6-406a-9600-4e7f2ff47f65)

(this is from the CRB 15000 GoFa 10 & 12 Datasheet (from [here](https://new.abb.com/products/robotics/robots/collaborative-robots/crb-15000)))

Same for joints 2 & 3.

Could you clarify where you found the current limits?

mmmarcopalma replied:
these values were simply not updated from the GoFa 5, my bad.

mmmarcopalma replied:
besides mesh normals, i realiize now link_4 is also not clsoed and has a bunch of non-manifold edges.  
fixing all of it right now.

gavanderhoorn commented on abb_crb15000_support/urdf/crb15000_12_127_macro.xacro:
```diff
@@ -0,0 +1,169 @@
+      <axis xyz="1 0 0"/>
+      <parent link="${prefix}link_5"/>
+      <child link="${prefix}link_6"/>
+      <limit effort="0" lower="${radians(-180)}" upper="${radians(180)}" velocity="${radians(200)}"/>
```
The position limits here seem to be different from the documentation I quoted earlier: `+-180` vs `+-270` degrees.

mmmarcopalma replied:
see previous comment.

gavanderhoorn reviewed (changes requested):
Thanks again for the PR. Holiday period got in the way.

See the inline comments for some things that need to be checked and possibly corrected.

Apart from those, I'm wondering whether there is an issue with the normals of `visual` meshes `link_1.stl` and `link_4.stl`:

![rviz_screenshot_2024_08_17-17_22_27](https://github.com/user-attachments/assets/e396bb06-f7d5-44a5-ba7b-c874bd79cb35)

they appear to be inverted, at least partially.

mmmarcopalma committed:
Unified mesh normals, fixed joint and speed limits.

mmmarcopalma commented:
Committed the requested changes. 
Thanks for supporting!

mmmarcopalma commented:
Hello @gavanderhoorn, is there anything else I can do to push this PR? 
Thanks!

gavanderhoorn committed:
Make sure to run 12/1.27 variant roslaunch test as well

gavanderhoorn committed:
Clarify source of joint limits and other specs

gavanderhoorn commented:
I've re-enabled CI and added a couple of commits.

I believe this is in an OK state to merge. Thanks again for the PR @mmmarcopalma :+1: 

Let's wait on CI to turn green.

gavanderhoorn commented:
To prevent all the fixups from ending up in the commit history, I'll squash-merge this.

Provenance and attribution will be retained, of course.

gavanderhoorn commented:
Thanks again @mmmarcopalma :+1:

mmmarcopalma commented:
Great news, thanks for the support @gavanderhoorn, and thanks also to @gonzalocasas for initiating the process
